File: ios/SaleMove/SaleMove.m
//
//  Salemove.m
//  ReactNativeDemo
//
//  Created by Siim Halapuu on 05/03/2018.
//  Copyright © 2018 Facebook. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>
#import <SalemoveSDK/SalemoveSDK-Swift.h>

@interface RCT_EXTERN_MODULE(SaleMove, NSObject)

RCT_EXTERN_METHOD(sendMessage:(NSString *)messageText callback:(RCTResponseSenderBlock)callback)

RCT_EXTERN_METHOD(listQueues:(RCTResponseSenderBlock)callback)

RCT_EXTERN_METHOD(requestEngagement:(NSString *)queueID withResolve:(RCTPromiseResolveBlock)resolve
                  withReject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(updateInformation:(NSString *)name
                  email:(NSString *)email
                  phone:(NSString *)phone
                  externalID:(NSString *)externalID
                  customAttrtibutes:(NSDictionary *)customAttributes
                  withResolve:(RCTPromiseResolveBlock)resolve
                  withReject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(endEngagement:(RCTPromiseResolveBlock)resolve
                  withReject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(requestTwoWayAudio:(RCTPromiseResolveBlock)resolve
                  withReject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(requestTwoWayVideo:(RCTPromiseResolveBlock)resolve
                  withReject:(RCTPromiseRejectBlock)reject)

@end

File: ios/SaleMove/SaleMove.swift
import Foundation

@objc(SaleMove)
class SaleMove: NSObject {
    @objc(requestTwoWayAudio:withReject:)
    func requestTwoWayAudio(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let offer = MediaUpgradeOffer(type: .audio, direction: .twoWay)
        Salemove.sharedInstance.requestMediaUpgrade(offer: offer) { success, error in
            guard success else {
                reject("", error?.reason, error?.error)
                return
            }
            
            resolve("")
        }
    }
    
    @objc(requestTwoWayVideo:withReject:)
    func requestTwoWayVideo(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let offer = MediaUpgradeOffer(type: .video, direction: .twoWay)
        Salemove.sharedInstance.requestMediaUpgrade(offer: offer) { success, error in
            guard success else {
                reject("", error?.reason, error?.error)
                return
            }
            
            resolve("")
        }
    }
    
    @objc(sendMessage:callback:)
    func sendMessage(messageText: String, callback: @escaping RCTResponseSenderBlock) {
        Salemove.sharedInstance.send(message: messageText) { message, error in
            let incomingError = error ?? NSNull()
            
            guard let incomingMessage = message else {
                callback([incomingError, NSNull()])
                return
            }
           
            let payload = ["id" : incomingMessage.id,
                           "content": incomingMessage.content,
                           "sender": "visitor"]
            
            callback([incomingError, payload])
        }
    }
    
    @objc(listQueues:)
    func listQueues(callback: @escaping RCTResponseSenderBlock) {
        Salemove.sharedInstance.listQueues { queues, error in
            let incomingError = error ?? NSNull()
            
            guard let incomingQueues = queues else {
                callback([incomingError, NSNull()])
                return
            }
            
            var payload: [[AnyHashable: Any]] = []
            
            for queue in incomingQueues {
                payload.append(["id": queue.id,
                                "name": queue.name])
            }
            
            callback([incomingError, payload])
        }
    }
    @objc(requestEngagement:withResolve:withReject:)
    func requestEngagement(queueID: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let context = VisitorContext(type: .page, url: "https://www.libertymutual.com")

        Salemove.sharedInstance.queueForEngagement(queueID: queueID, visitorContext: context) { queueTicket, error in
            guard let incomingTicket = queueTicket else {
                reject("", error?.reason, error?.error)
                return
            }
            
            resolve("")
        }
    }
    
    @objc(endEngagement:withReject:)
    func endEngagement(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        Salemove.sharedInstance.endEngagement { success, error in
            guard success else {
                reject("", error?.reason, error?.error)
                return
            }
            
            resolve("")
        }
    }
}
